File: codes/cpp/chap03/answer_A13.cpp
#include <iostream>
using namespace std;

int N, K;
int A[100009], R[100009];

int main() {
	// 入力
	cin >> N >> K;
	for (int i = 1; i <= N; i++) cin >> A[i];

	// しゃくとり法
	for (int i = 1; i <= N - 1; i++) {
		// スタート地点を決める
		if (i == 1) R[i] = 1;
		else R[i] = R[i - 1];

		// ギリギリまで増やしていく
		while (R[i] < N && A[R[i] + 1] - A[i] <= K) {
			R[i] += 1;
		}
	}

	// 出力（答えは最大 50 億程度になるので long long 型を使う必要があります）
	long long Answer = 0;
	for (int i = 1; i <= N - 1; i++) Answer += (R[i] - i);
	cout << Answer << endl;
	return 0;
}

File: codes/java/chap03/answer_A13.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		// 入力
		int N = sc.nextInt();
		int K = sc.nextInt();
		int[] A = new int[N + 1];
		for (int i = 1; i <= N; i++) A[i] = sc.nextInt();

		// 配列 R の定義
		int[] R = new int[N + 1];

		// しゃくとり法
		for (int i = 1; i <= N - 1; i++) {
			// スタート地点を決める
			if (i == 1) R[i] = 1;
			else R[i] = R[i - 1];

			// ギリギリまで増やしていく
			while (R[i] < N && A[R[i] + 1] - A[i] <= K) {
				R[i] += 1;
			}
		}

		// 出力（答えは最大 50 億程度になるので long 型を使う必要があります）
		long Answer = 0;
		for (int i = 1; i <= N - 1; i++) Answer += (R[i] - i);
		System.out.println(Answer);
	}
};
